### PR TITLE
CELDEV-266: Add NavigationFactory allow different configuration sources

### PR DESCRIPTION
**Pull request by fpichler**
https://synjira.atlassian.net/browse/CELDEV-266

### CODE REVIEW

File: src/main/java/com/celements/navigation/Navigation.java [hunk 1]
```diff
@@ -77,13 +81,12 @@
 
   private INavigationBuilder navBuilder;
 
-  public static final String LIST_LAYOUT_TYPE = "list";
-
-  private static final String _PAGE_MENU_DATA_TYPE = "pageMenu";
+  @Deprecated
+  public static final String LIST_LAYOUT_TYPE = NavigationConfig.LIST_LAYOUT_TYPE;
 
   public static final String MENU_TYPE_MENUITEM = "menuitem";
```
Comment (ebeutler) [inline src/main/java/com/celements/navigation/Navigation.java]: Is there a reason MENU_TYPE_MENUITEM was not moved to NavigationConfig like all the other constants?

Comment (fpichler) [inline src/main/java/com/celements/navigation/Navigation.java]: MENU_TYPE... is not a ClassConfig name. It is a Menu Type name.

**Comment by ebeutler** [comment]
looks fine to me, but maybe @msladek wants to have a look too.

File: src/main/java/com/celements/navigation/NavigationConfig.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+  }
+
+  @Nullable
+  final private <T> T firstNotNullOrNull(@Nullable T firstObj, @Nullable T secondObj) {
+    if (firstObj != null) {
+      return firstObj;
+    } else {
+      return secondObj;
+    }
+  }
```
Comment (msladek) [inline src/main/java/com/celements/navigation/NavigationConfig.java]: This method can be replaced with `org.apache.commons.lang3.ObjectUtils.firstNonNull(T...)`

**Comment by fpichler** [comment]
I refactored completely implementing the builder pattern and guava Optionals. Please review again.

**Comment by ebeutler** [comment]
The changes look fine.

File: src/main/java/com/celements/navigation/Navigation.java [hunk 3]
```diff
@@ -77,13 +81,12 @@
...
 
-  private static Log LOGGER = LogFactory.getFactory().getInstance(Navigation.class);
+  private final static Logger LOGGER = LoggerFactory.getLogger(Navigation.class);
 
   private static final String _NAVIGATION_COUNTER_KEY = NavigationApi.class.getCanonicalName()
       + "_NavigationCounter";
```
Comment (msladek) [inline src/main/java/com/celements/navigation/Navigation.java]: variables names starting with underscore not allowed according to Google Java Style Guide

File: src/main/java/com/celements/navigation/NavigationConfig.java [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+  public static final int DEFAULT_MAX_LEVEL = 100;
+  public static final String PAGE_MENU_DATA_TYPE = "pageMenu";
+  public static final String LIST_LAYOUT_TYPE = "list";
+
+  public static class Builder {
+
+    // Required parameters
+    private boolean enabled;
+
+    // Optional parameters - initialized to Optional
```
Comment (msladek) [inline src/main/java/com/celements/navigation/NavigationConfig.java]: initialized to absent?

File: src/main/java/com/celements/navigation/NavigationConfig.java [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+    this.layoutType = builder.layoutType;
+    this.nrOfItemsPerPage = builder.nrOfItemsPerPage;
+    this.presentationTypeHint = builder.presentationTypeHint;
+    this.cmCssClass = builder.cmCssClass;
+  }
+
+  @Nullable
+  final private <T> T firstNotNullOrNull(@Nullable T firstObj, @Nullable T secondObj) {
+    return Optional.fromNullable(firstObj).or(Optional.fromNullable(secondObj)).orNull();
+  }
```
Comment (msladek) [inline src/main/java/com/celements/navigation/NavigationConfig.java]: this method is never used

File: src/main/java/com/celements/navigation/NavigationConfig.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+
+  @NotNull
+  public String getMenuPart() {
+    return menuPart.or("");
+  }
+
+  @Nullable
+  public SpaceReference getNodeSpaceRef() {
+    if (nodeSpaceRef.isPresent()) {
+      return new SpaceReference(nodeSpaceRef.get().clone());
```
Comment (msladek) [inline src/main/java/com/celements/navigation/NavigationConfig.java]: please use ModelUtils.cloneReference(...)

**Comment by msladek** [comment]
looks good now apart from minor comments :)

**Pull request merged**